hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/FloatConverter.java

```diff
@@ -38,27 +40,58 @@ private FloatConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        float casted = cast(val);
+        byte converted = (byte) casted;
+
+        if (converted != (int) casted || !Float.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.TINYINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public short asSmallint(Object val) {
-        return (short) cast(val);
+        float casted = cast(val);
+        short converted = (short) casted;
+
+        if (converted != (int) casted || !Float.isFinite(casted)) {
```

Comment:
Why do we cast to `int` here to check for overflow, but cast to `long` in the `asInt` method? 
Is it possible that due to overflow `castedInt != castedLong` but `castedInt == castedShort`?

Reply:
When `float` is casted to `int`, the so called saturating arithmetic cast is performed: any `float` value outside of the `int` representable range is converted to either `Integer.MIN_VALUE` or `Integer.MAX_VALUE`. It's obvious that any `short` representable value is within that range and can be represented without magnitude loss.

When `float` is casted to `short`, it is first casted to `int`, because `short` is not a first class citizen in Java bytecode, then the produced `int` is implicitly casted to `short`, that's the time when the overflow may happen. If the `short` value implicitly casted back to `int` doesn't match the `int` value, that means the overflow happened.

Similar logic applies to `long` and `int` combo. See Narrowing Primitive Conversion section of JLS for more details.

Reply:
Thanks for the explanation. I would add it to JavaDoc, because otherwise, it is hard to judge about the code correctness.

Reply:
fixed